Add in basic tag/category search functionality

abethcrane opened the pull request:
Tag and category pages can now be automatically generated by using the search page, e.g.
"/search.html?tags=baking".

Currently only supports single tag/category queries, combinations is left as future work.

All posts have been updated to have simple tags.

Commit:
Updating for hosting on github pages so I can test things

Commit:
Updating base url to end with slash

Commit:
Fix up recipe_list layout to prepend baseurl to post.url

Commit:
Merge branch 'master' into gh-pages

Commit:
whitespace

Commit:
whitespace

Commit:
Fix jekyll bug of base_url ending with slash and permalink starting with one

Commit:
Finish getting it to run on github pages

Commit:
Adding tags to icecream cupcakes for some testing

Commit:
Revert "Adding tags to icecream cupcakes for some testing"

This reverts commit 2f6895e37de810c56ac71a57c38daedcc2d5c50c.

Commit:
Adding tags to icecream cupcakes for some testing

Commit:
Tags on newlines

Commit:
Tags space-separated with scroll

Commit:
Proper horizontal scrolling for tags

Commit:
Comma-separated tags

Commit:
Initial dump of stuff for dynamically generated tags/categories pages

Commit:
Typo in link

Commit:
Fix up search links misspelling

Commit:
More link tweaking - prepend the base url

Commit:
Fixing up link to use proper templating syntax

Commit:
Adding a bunch of files for searching

Commit:
Alert testing for search functionality

Commit:
Adding search.js to footer - all pages and loading at end of page

Commit:
Cleaning up search.html

Commit:
Typo: Missing quote

Commit:
Add Jquery

Commit:
update jquery link

Commit:
Update js to have open-sauce in path

Commit:
Attempting to get search.json properly populated

Commit:
Attempting to get search.json properly populated part 2

Commit:
Attempting to get search.json properly populated part 3

Commit:
Change default layout to wrapper

default needs to be blank for json generation

Commit:
Remove recipe as default layout so that json will populate

Commit:
Switch back from wrapper to default

Commit:
Attempting to get search.json properly populated part 4

Commit:
Attempting to get search.json properly populated part 5

Commit:
Typo: removed dash in config

Commit:
Adding in handlebars

Commit:
Search.json disappeared :(

Commit:
Search.json disappeared. Debugging 1

Commit:
Search.json disappeared. Debugging 2 - uninclude handlebar

Commit:
Fix up config file error

Commit:
Set up of search

- Include handlebars and search page template
- Properly populate json site index
- .gitignore gem files for local jekyll hosting
- Add scripts to footer

Commit:
Fixing up last few search tags things

Commit:
Update some more recipes to have tags

Commit:
Adding tags

Commit:
Fixing up tabs/spaces

Commit:
Minor improvements to search page

Commit:
Merge branch 'gh-pages'

Commit:
Set it back to baseurl of / rather than /open-sauce/

Commit:
whitespace

Commit:
Last post to tag!

tetrakai commented on _config.yml:
```diff
@@ -14,4 +14,4 @@ highlighter: pygments
   -
     values:
-      layout: recipe
+      layout: recipe
```
Nit: could you add a newline to the end of this file?

tetrakai commented on .gitignore:
```diff
@@ -1,2 +1,6 @@
 .sass-cache
+
+Gemfile
+
```
Nit: getting rid of the newlines here would be cleaner

tetrakai commented on _includes/post.json:
```diff
@@ -0,0 +1,8 @@
+{
+  "title"    : "{{ post.title }}",
+  "category" : "{{ post.category }}",
+  "tags"     : [{% for tag in post.tags %}"{{ tag }}",{% endfor %} null],
```
Is there a reason to always add a null tag here?

abethcrane replied:
Nope, it was instead of having a check for unless last to deal with commas. Fixed.

tetrakai commented on _includes/post.json:
```diff
@@ -0,0 +1,8 @@
+  "url"     : "{{ post.url }}",
+  "title_image" : "{{ post.title_image }}",
+  "prep_time" : "{{ post.prep_time }}"
+}
```
We might want to think about the best way to escape these properties - if someone creates a recipe with a title containing double quotes, this will break

abethcrane replied:
"{{ post.title | xml_escape }}" seems like an easy way to do that for generating the json. I'm not sure the nicest way to convert it back to html for displaying - if the search_results_list could be parsed with jekyll with markdownify, but handlebars helpers are more work.

abethcrane replied:
turns out {{ | jsonify }} is nice for this :)

tetrakai commented on _layouts/empty.html:
```diff
@@ -0,0 +1 @@
+{{ content }}
```
This file seems a bit redundant - I *think* it's possible to not specify a layout in Jekyll, though I could be wrong

tetrakai replied:
If we do keep it, adding a newline to the end of this file would also be nice

abethcrane replied:
It's possible to not specify a layout, but I was having trouble with getting the different rules to work out - you have the default as recipes. I think we could change that to default for posts, but it wasn't playing nice.

tetrakai commented on _layouts/recipe.html:
```diff
@@ -46,10 +46,21 @@
-              </dl>
+                <dd><a href = "{{ page.category | downcase | prepend: 'search.html?category=' | prepend: site.baseurl }}">{{ page.category }}</a></dd>
+                {% capture difference %} {{ page.tags | size | minus:1 }} {% endcapture %}
+                {% unless difference contains '-' %}
```
I think `{% if page.tags.size > 0 %}` would be cleaner here

tetrakai commented on _layouts/recipe.html:
```diff
@@ -46,10 +46,21 @@
+                </dl>
 
-              <dl class="ingredients">
+                <dl class="ingredients">
```
Why was this indented?

tetrakai commented on _layouts/recipe.html:
```diff
@@ -46,10 +46,21 @@
+                  {% assign sorted_tags = (page.tags | sort) %}
+                  {% for tag in sorted_tags %}
+                <a href="{{ tag | downcase | prepend: 'search.html?tags=' | prepend: site.baseurl }}">{{ tag }}</a>
+                    {% if forloop.last == false %}, {% endif %}
```
`{% if not forloop.last %}`, or `{% unless forloop.last %}` would be clearer here

tetrakai commented on js/handlebar.js:
```diff
@@ -0,0 +1,2746 @@
+
+      return this.createFunctionContext(asObject);
+    },
+
```
Could we use a minified version of this file instead? I don't imagine we'll be modifying it in future, and it will make this diff look like less of a behemoth

tetrakai commented on js/search.js:
```diff
@@ -0,0 +1,122 @@
+  posts.pop();
+  for (var i in posts) {
+    var post = posts[i],
+    prop = post[property];
```
Is this variable intentionally global?

tetrakai commented on js/search.js:
```diff
@@ -0,0 +1,122 @@
+  dict["posts"] = posts;
+  dict["baseurl"] = "";
+
+  console.log("Going to display: ", dict);
```
Probably better to get rid of this :)

tetrakai commented on js/search.js:
```diff
@@ -0,0 +1,122 @@
+// Returns:
+//  undefined
+var noResultsPage = function(property, value) {
+  console.log ("No results found");
```
It'd be nice to get rid of this too

tetrakai commented on js/search.js:
```diff
@@ -0,0 +1,122 @@
+    '<br/><br/>\
+    <h1>No Results Found</h1>\
+    <p>Sorry, none of these recipes are described as ‘' + value + '’.</p>'
+  );
```
This introduces a reflected XSS vulnerability - the user can control `value` and set it to `<img src='/nothing' onerror='alert("Running JS")' />`. While this isn't severe at the moment, since the site doesn't have sessions, it would be better to fix it.

Putting this HTML into the page by default, in a hidden div, and showing that div when no results are found, is a safer way of doing this. You can set the value you want on the page by adding a span, then referencing that span by ID and setting it's .text (which is sanitized, where .html isn't).

This might look something like this:
`$('#results').show();
$('#results .value').text(value);`

tetrakai commented on js/search.js:
```diff
@@ -0,0 +1,122 @@
+
+  $.each(map, function(type, value) {
+    if (value !== null) {
+      $.getJSON('/search.json', function(data) { //TODO: not hardcode the base url
```
Could you fix the TODO here before we merge this? :)

tetrakai commented on search.json:
```diff
@@ -0,0 +1,9 @@
+    {% include post.json %},
+  {% endfor %}
+  null
+]
```
Nit: Adding a newline to the end of this file would be awesome

tetrakai commented on search_results_list.html:
```diff
@@ -0,0 +1,34 @@
+            <div class="entry one-third" style="margin-bottom: 15px">
+              <!-- Recipe picture -->
+              <figure style="height: 169px; overflow: hidden;">
+                <img src="{{ ../baseurl }}/img/recipes/{{ this.title_image }}" alt="{{ this.title }}" />
```
It'd be nice to do this the way it's done elsewhere - `{{ this.title_image | prepend: 'img/recipes'  | prepend: site.baseurl }}`, though we're now doing this in 3 places and should consider adding it to the whole-site config

abethcrane replied:
Ah, this file is not actually jekyll, it's handlebars. I'd like to find a way to make it the same, but for now the syntax is slightly different.

tetrakai commented on search_results_list.html:
```diff
@@ -0,0 +1,34 @@
+              <!-- Recipe picture -->
+              <figure style="height: 169px; overflow: hidden;">
+                <img src="{{ ../baseurl }}/img/recipes/{{ this.title_image }}" alt="{{ this.title }}" />
+                <figcaption><a href="{{ ../baseurl }}{{ this.url }}"><i class="ico i-view"></i> <span>View recipe</span></a></figcaption>
```
Ditto the above comment :)

tetrakai commented:
This is awesome overall, thanks!
There are a couple of minor things to clean up before merging, but the main things are removing the XSS vulnerability and smoothing over the URLs.

Commit:
Specify default empty layout

https://github.com/tetrakai/open-sauce/pull/2#discussion_r28199754

https://github.com/tetrakai/open-sauce/pull/2#discussion-diff-28199802

Commit:
post.json - remove null and add jsonify

https://github.com/tetrakai/open-sauce/pull/2#discussion_r28199763

https://github.com/tetrakai/open-sauce/pull/2#discussion_r28199782

Commit:
Remove empty.html

https://github.com/tetrakai/open-sauce/pull/2#discussion_r28199802

Commit:
Clean up page tags and indentation

https://github.com/tetrakai/open-sauce/pull/2#discussion_r28199981

https://github.com/tetrakai/open-sauce/pull/2#discussion_r28199982

https://github.com/tetrakai/open-sauce/pull/2#discussion_r28199998

Commit:
gitignore nit

https://github.com/tetrakai/open-sauce/pull/2#discussion_r28199760

Commit:
Minify handlebars

https://github.com/tetrakai/open-sauce/pull/2#discussion_r28200002

Commit:
Search.js cleanup - remove console logs, fix xss vulnerability

https://github.com/tetrakai/open-sauce/pull/2#discussion_r28200008

https://github.com/tetrakai/open-sauce/pull/2#discussion_r28200012

https://github.com/tetrakai/open-sauce/pull/2#discussion_r28200018

https://github.com/tetrakai/open-sauce/pull/2#discussion_r28200054

https://github.com/tetrakai/open-sauce/pull/2#discussion_r28200057

Commit:
Clean up search.json - commas and newline

https://github.com/tetrakai/open-sauce/pull/2#discussion_r28200059

Commit:
search.html fix up xss vuln by including hidden results div
